numbergen - DeprecationWarning: Seeding based on hashing is deprecated since Python 3.9

From Python 3.9 onward a deprecation warning is emitted if a seed is set from a tuple, as done here in `numbergen`: https://github.com/holoviz/param/blob/32060b24736fa732417103a12ea05f941cf1f660/numbergen/__init__.py#L310

```
 _________________ ERROR collecting tests/API1/testnumbergen.py _________________
tests/API1/testnumbergen.py:4: in <module>
    import numbergen
numbergen/__init__.py:277: in <module>
    class TimeAwareRandomState(TimeAware):
numbergen/__init__.py:310: in TimeAwareRandomState
    default=random.Random((500,500)), doc=
/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/random.py:123: in __init__
    self.seed(x)
/opt/hostedtoolcache/Python/3.9.10/x64/lib/python3.9/random.py:158: in seed
    _warn('Seeding based on hashing is deprecated\n'
E   DeprecationWarning: Seeding based on hashing is deprecated
E   since Python 3.9 and will be removed in a subsequent version. The only 
E   supported seed types are: None, int, float, str, bytes, and bytearray.
```

@jlstevens had some idea about how to fix that. Note that if the warning is eventually turned into an error in future Python versions `numbergen` will no longer be importable, since this `randon.Random((500, 500))` is executed on import time.

The deprecation turns into an error in Python 3.11, which is currently in beta. I’m the maintainer of the [`python-param` package](https://src.fedoraproject.org/rpms/python-param) in Fedora Linux, and we’re moving to Python 3.11 in our development version (Rawhide) now, so this is [already broken for us](https://bugzilla.redhat.com/show_bug.cgi?id=2093926), with errors like:

```
_______________ ERROR collecting tests/API0/testdynamicparams.py _______________
tests/API0/testdynamicparams.py:14: in <module>
    import numbergen
numbergen/__init__.py:277: in <module>
    class TimeAwareRandomState(TimeAware):
numbergen/__init__.py:310: in TimeAwareRandomState
    default=random.Random((500,500)), doc=
/usr/lib64/python3.11/random.py:125: in __init__
    self.seed(x)
/usr/lib64/python3.11/random.py:160: in seed
    raise TypeError('The only supported seed types are: None,\n'
E   TypeError: The only supported seed types are: None,
E   int, float, str, bytes, and bytearray.
```

To understand how `(500, 500)` becomes a RNG state, we can check the CPython source code [in Modules/_randommodule.c](https://github.com/python/cpython/blob/f9585e2adc36fb2886f96ef86a5aa14215fc1151/Modules/_randommodule.c#L299):

```c
static PyObject *
random_seed(RandomObject *self, PyObject *arg)
{
    PyObject *result = NULL;            /* guilty until proved innocent */
    PyObject *n = NULL;
    /* [... snip ... ] */

    if (arg == NULL || arg == Py_None) {
        /* [... snip ... ] */
    }

    /* This algorithm relies on the number being unsigned.
     * So: if the arg is a PyLong, use its absolute value.
     * Otherwise use its hash value, cast to unsigned.
     */
    if (PyLong_CheckExact(arg)) {
        n = PyNumber_Absolute(arg);
    } else if (PyLong_Check(arg)) {
        /* [... snip ... ] */
    }
    else {
        Py_hash_t hash = PyObject_Hash(arg);
        if (hash == -1)
            goto Done;
        n = PyLong_FromSize_t((size_t)hash);
    }
    /* [... snip ... ] */
}
```

Thus for a non-numeric seed like this, the actual seed is its Python `hash()`, cast C-style to a `size_t`. That is:

```python
from random import Random
from ctypes import c_size_t

seed = (500, 500)
a = Random(seed)
b = Random(c_size_t(hash(seed)).value)

assert a.getstate() == b.getstate()
assert a.random() == b.random()
```

So we could explicitly continue the current behavior on Python 3.11 by writing:

```python
from ctypes import c_size_t
```

and then

```python
        default=random.Random(c_size_t(hash((500,500))).value), doc=
```

-----

As a side note, a big part of the reason that using arbitrary objects as seeds was removed is that [string hashes are randomized](https://peps.python.org/pep-0456/) in Python, so using hashable objects that contain strings can lead to surprising behavior, e.g.:

```bash
$ python3.10 -c 'from random import Random; print(Random(("500", "500")).random())'
0.543252271025492
$ python3.10 -c 'from random import Random; print(Random(("500", "500")).random())'
0.48946932166185464
```

The hash of a tuple-of-numbers like `(500, 500)` is still repeatable, though.

-----

One caveat about the way `Random` is seeded here is that the result will be, and *already is*, different on 64-bit and 32-bit platforms.

**64-bit**:

```
$ python3.10 -c 'from random import Random; print(Random((500, 500)).random())'
0.1248426853025093
```

**32-bit**:

```
$ python3.10 -c 'from random import Random; print(Random((500, 500)).random())'
0.7817896725813163
```

It’s also different between newer and older Pythons, as the hash algorithm has changed:

```
$ python2.7 -c 'from random import Random; print(Random((500, 500)).random())'
0.22241681053
```

So if we are being strict about preserving the same behavior, `c_size_t(hash((500,500))).value` cannot be simplified at all. On the other hand, if you wrote `13594734499543488215`, then all the other cases (Python 2.7, 32-bit builds) would change to match what you currently see on 64-bit builds of supported versions of Python 3.

-----

What do you think? I’m happy to follow up with a PR.

--- a/numbergen/__init__.py
+++ b/numbergen/__init__.py
@@ -8,6 +8,7 @@ import hashlib
 import struct
 import fractions
 
+from ctypes import c_size_t
 from math import e,pi
 
 import param
@@ -306,8 +307,18 @@ class TimeAwareRandomState(TimeAware):
     explicitly when you construct the RandomDistribution object.
     """
 
+    # Historically, the default random state was seeded with the tuple
+    # (500, 500). The CPython implementation implicitly formed an unsigned
+    # integer seed using the hash of the tuple as in the expression below. Note
+    # that the resulting integer, and therefore the default initial random
+    # state, varies across CPython versions (as the hash algorithm has changed)
+    # and also between 32-bit and 64-bit interpreters.
+    #
+    # Seeding based on hashing is deprecated since Python 3.9 and removed in
+    # Python 3.11; we explicitly continue the historical behavior for the time
+    # being.
     random_generator = param.Parameter(
-        default=random.Random((500,500)), doc=
+        default=random.Random(c_size_t(hash((500,500))).value), doc=
         """
         Random state used by the object. This may may be an instance
         of random.Random from the Python standard library or an

